connectors/sources/google_drive.py

```diff
@@ -558,100 +888,167 @@ async def get_content(self, blob, timestamp=None, doit=None):
+        last_modifying_user = file.get("lastModifyingUser", None)
         if last_modifying_user:
-            blob_document["updated_by"] = last_modifying_user.get("displayName", None)
-            blob_document["updated_by_email"] = last_modifying_user.get(
+            file_document["updated_by"] = last_modifying_user.get("displayName", None)
+            file_document["updated_by_email"] = last_modifying_user.get(
                 "emailAddress", None
             )
-            blob_document["updated_by_photo_url"] = last_modifying_user.get(
+            file_document["updated_by_photo_url"] = last_modifying_user.get(
                 "photoLink", None
             )
 
         # determine the path on google drive, note that google workspace files won't have a path
-        blob_parents = blob.get("parents", None)
-        if blob_parents and blob_parents[0] in paths:
-            blob_document["path"] = f"{paths[blob_parents[0]]['path']}/{blob['name']}"
+        file_parents = file.get("parents", None)
+        if file_parents and file_parents[0] in paths:
+            file_document["path"] = f"{paths[file_parents[0]]['path']}/{file['name']}"
+
+        # mark the document if it is on shared drive
+        file_drive_id = file.get("driveId", None)
+        shared_drive = paths.get(file_drive_id, None)
+        if shared_drive:
+            file_document["shared_drive"] = shared_drive.get("name")
+
+        if self._dls_enabled():
+            # Getting permissions works differenty for files on my drive and files on shared drives.
+            # Read more: https://developers.google.com/drive/api/guides/shared-drives-diffs
+            permissions = file.get("permissions", [])
+            if not permissions:
+                permissions = await self._get_permissions_on_shared_drive(
+                    file_id=file_id
+                )
+
+            file_document[ACCESS_CONTROL] = self._process_permissions(permissions)
 
-        return blob_document
+        return file_document
 
-    def get_blob_document(self, blobs, paths):
-        """Generate blob document.
+    async def prepare_files(self, files_page, paths):
+        """Generate file document.
 
         Args:
-            blobs (dict): Dictionary contains blobs list.
+            files_page (dict): Dictionary contains files list.
 
         Yields:
-            dict: Blobs metadata mapped with the keys of `BLOB_ADAPTER`.
+            dict: File with formatted metadata.
         """
-        for blob in blobs.get("files", []):
-            yield self.prepare_blob_document(blob=blob, paths=paths)
+        files = files_page.get("files", [])
+
+        async def process_file(file, semaphore):
+            async with semaphore:
+                return await self.prepare_file(file=file, paths=paths)
+
+        # Create the shared semaphore, it controls how many concurrent
+        # permissions/list requests can be open at any given time
+        semaphore = asyncio.Semaphore(20)
```

Comment:
question: Looks similar to the logic processing users. Can we DRY something up here?

Reply:
Yup, extracted the common logic to: `_process_items_concurrently`